UM/Trust.py

```diff
@@ -0,0 +1,214 @@
+        return False
+
+
+# Trust as a singleton class for the main code, as opposed to the (keygen/signing) scripts:
+class Trust:
+    __instance = None
+
+    @staticmethod
+    def getPublicRootKeyPath():
+        from UM.Application import Application
+        return os.path.abspath(os.path.join(Application.getAppFolderPrefix(), "public_key.pem"))
+
+    @classmethod
+    def getInstance(cls):
+        if not cls.__instance:
+            cls.__instance = Trust(Trust.getPublicRootKeyPath())
+        return cls.__instance if cls.__instance._public_key else None
+
+    def __init__(self, public_key_filename: str) -> None:
+        self._public_key = None  #type: Optional[RSAPublicKey]
+        try:
+            with open(public_key_filename, "rb") as file:
+                self._public_key = load_pem_public_key(file.read(), backend = default_backend())
+        except:  # Yes, we  do really want this on _every_ exception that might occur.
+            Logger.logException("e", "Couldn't load public-key '{0}'.".format(public_key_filename))
+            self._public_key = None
+
+    def _verifyFile(self, filename: str, signature: str) -> bool:
+        if self._public_key is None:
+            return False
+        file_hash = TrustBasics.getFileHash(filename)
+        if file_hash == "":
+            return False
+        try:
+            signature_bytes = base64.b64decode(signature)
+            file_hash_bytes = base64.b64decode(file_hash)
+            self._public_key.verify(
+                signature_bytes,
+                file_hash_bytes,
+                padding.PSS(mgf = padding.MGF1(TrustBasics.getHashAlgorithm()), salt_length = padding.PSS.MAX_LENGTH),
+                Prehashed(TrustBasics.getHashAlgorithm())
+            )
+            return True
+        except:  # Yes, we  do really want this on _every_ exception that might occur.
+            Logger.logException("e", "Couldn't verify '{0}' with supplied signature.".format(filename))
+        return False
+
+    def signedFolderCheck(self, path: str) -> bool:
+        try:
+            json_filename = os.path.join(path, TrustBasics.getSignaturesLocalFilename())
+
+            with open(json_filename, "r", encoding = "utf-8") as data_file:
+                signatures_json = json.load(data_file)
+
+                file_count = 0
+                for root, dirnames, filenames in os.walk(path):
+                    for filename in filenames:
+                        if filename == TrustBasics.getSignaturesLocalFilename() and root == path:
+                            continue
+                        file_count += 1
+                        name_on_disk = os.path.join(root, filename)
+                        name_in_data = name_on_disk.replace(path, "").replace("\\", "/")
+
+                        signature = signatures_json.get(name_in_data, None)
+                        if signature is None:
```

Comment:
The file-hash won't be empty, _even_ in case the file _is_. In the case a file-hash _is_ somehow empty, it still won't match (even an empty itself) signature.